napari_animation/_qt/keyframeslist_widget.py

```diff
@@ -1,10 +1,214 @@
+
+        self.animation = animation
+        self._frame_counter = 0
+        self._key_frame_id_to_label_map = {}
+        self._label_to_qlistwidgetitem_map = {}
+
+        self._connect_key_frame_callbacks()
+        self.setDragDropMode(super().InternalMove)
+        self.setIconSize(QSize(64, 64))
+
+        self.itemClicked.connect(self._selection_callback)
+
+    def _connect_key_frame_callbacks(self):
+        """Connect events on the key frame list to their callbacks
+        """
+        self.animation.key_frames.events.removed.connect(self._remove)
+        self.animation.key_frames.events.moved.connect(self._update_frontend)
+        self.animation.key_frames.events.changed.connect(self._update_frontend)
+        self.animation.key_frames.events.reordered.connect(self._reorder)
+
+    def dropEvent(self, event):
+        """update animation state on 'drop' of frame in key frames list
+        """
+        super().dropEvent(event)
+        self._update_backend()
+
+    def _selection_callback(self, event):
+        self._update_frame_number()
+        self.animation.set_to_current_keyframe()
+        self.parentWidget()._update_frame_widget_from_animation()
+
+    def _add(self):
+        """Generate QListWidgetItem for current keyframe, store its unique id and add it to self
+        """
+        key_frame_idx = self.animation.frame
+        item = self._generate_list_item(key_frame_idx)
+
+        self.insertItem(key_frame_idx, item)
+
+        self._map_key_frame_to_id(key_frame_idx)
+        self._map_label_to_qlistwidgetitem(key_frame_idx)
+        self._frame_counter += 1
+
+    def insertItem(self, row, item):
+        """overrides QListWidget.insertItem to also update current index and frame number
+        """
+        super().insertItem(row, item)
+        self.setCurrentIndex(self.indexFromItem(item))
+        self._update_frame_number()
+
+    def _remove(self, event):
+        """Remove QListWidgetItem at event.index
+        """
+        self.takeItem(event.index)
+        self._remove_key_frame_id(event.value)
+        self._update_frame_number()
+
+    def _remove_key_frame_id(self, key_frame):
+        """Remove key-frame id from self._key_frame_id_to_label_map
+        """
+        self._key_frame_id_to_label_map.pop(id(key_frame))
+
+    def _reorder(self, event):
+        """Reorder QListWidgetItems
+        """
+        self.clear()
+        for idx, _ in enumerate(self.animation.key_frames):
+            self.addItem(self._generate_list_item(idx))
+
+    def _update_backend(self):
+        """push current GUI state to self.animation
+        """
+        new_key_frames = [self._label_to_key_frame(label) for label in
+                          self.frontend_key_frame_labels]
+
+        # recreating the EventedList here and reconnecting events simplifies handling events
+        # which fire when we modify the list in place
+        self.animation.key_frames = EventedList(new_key_frames)
```

Comment:
Hmm yeah, this seems undesirable - it's not quite clear to me why we have to replace the whole list and we can't just reorder the elements and get events from this

Reply:
totally agree - it felt wrong, but I'll make sorting it a priority moving forward! :+1:

Reply:
See notes below, I'm happy to merge this as is, it's a big improvement, and then you can work on this in a followup PR